# import libraries

In [1]:
import numpy as np
import pandas as pd

# load data

### Data Columns

•	**id**: is the engine ID, ranging from 1 to 100  
•	**cycle**: per engine sequence, starts from 1 to the cycle number where failure had happened (training data only) 
•	__setting1__ to __setting3__: engine operational settings  
•	__s1__ to __s21__: sensors measurements  


## load training dataset

In [21]:
df_train = pd.read_csv('./data/PM_train.txt', sep=' ', header=None)
df_train.head()

0   1       2       3      4       5       6        7        8      9   \
0   1   1 -0.0007 -0.0004  100.0  518.67  641.82  1589.70  1400.60  14.62   
1   1   2  0.0019 -0.0003  100.0  518.67  642.15  1591.82  1403.14  14.62   
2   1   3 -0.0043  0.0003  100.0  518.67  642.35  1587.99  1404.20  14.62   
3   1   4  0.0007  0.0000  100.0  518.67  642.35  1582.79  1401.87  14.62   
4   1   5 -0.0019 -0.0002  100.0  518.67  642.37  1582.85  1406.22  14.62   

   ...       18      19    20   21    22     23     24       25  26  27  
0  ...  8138.62  8.4195  0.03  392  2388  100.0  39.06  23.4190 NaN NaN  
1  ...  8131.49  8.4318  0.03  392  2388  100.0  39.00  23.4236 NaN NaN  
2  ...  8133.23  8.4178  0.03  390  2388  100.0  38.95  23.3442 NaN NaN  
3  ...  8133.83  8.3682  0.03  392  2388  100.0  38.88  23.3739 NaN NaN  
4  ...  8133.80  8.4294  0.03  393  2388  100.0  38.90  23.4044 NaN NaN  

[5 rows x 28 columns]

In [22]:
df_feature = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 's1', 's2', 's3', 's4', 's5', 's6', 's7',
            's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20', 's21']

In [23]:
def prepare_df(df):
        """
        df : dataframe input to prepare for wrangling
        returns : dataframe

        """

        data = df
        df_feature = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 's1', 's2', 's3', 's4', 's5', 's6', 's7',
                's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20', 's21']

        data.drop(columns=[26,27], axis=1, inplace=True)
        data.columns = df_feature

        return data

In [24]:
df_train = prepare_df(df_train)
df_train.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...     s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62  ...  521.66  2388.02  8138.62  8.4195  0.03  392  2388  100.0   
1  14.62  ...  522.28  2388.07  8131.49  8.4318  0.03  392  2388  100.0   
2  14.62  ...  522.42  2388.03  8133.23  8.4178  0.03  390  2388  100.0   
3  14.62  ...  522.86  2388.08  8133.83  8.3682  0.03  392  2388  100.0   
4  14.62  ...  522.19  2388.04  8133.80  8.4294  0.03  393  2388  100.0   

     s20      s21  
0  39.06  23.4190  
1  39.00  23.4236  
2  38.95  23.3442  
3  38.88  23.3739  
4  38.90  23.4044  

[5 rows x 26 columns]

In [25]:
df_train.isna().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

In [26]:
df_train.describe()

id         cycle      setting1      setting2  setting3  \
count  20631.000000  20631.000000  20631.000000  20631.000000   20631.0   
mean      51.506568    108.807862     -0.000009      0.000002     100.0   
std       29.227633     68.880990      0.002187      0.000293       0.0   
min        1.000000      1.000000     -0.008700     -0.000600     100.0   
25%       26.000000     52.000000     -0.001500     -0.000200     100.0   
50%       52.000000    104.000000      0.000000      0.000000     100.0   
75%       77.000000    156.000000      0.001500      0.000300     100.0   
max      100.000000    362.000000      0.008700      0.000600     100.0   

                 s1            s2            s3            s4            s5  \
count  2.063100e+04  20631.000000  20631.000000  20631.000000  2.063100e+04   
mean   5.186700e+02    642.680934   1590.523119   1408.933782  1.462000e+01   
std    6.537152e-11      0.500053      6.131150      9.000605  3.394700e-12   
min    5.186700e+02    641.210000   1571.040000   1382.250000  1.462000e+01   
25%    5.186700e+02    642.325000   1586.260000   1402.360000  1.462000e+01   
50%    5.186700e+02    642.640000   1590.100000   1408.040000  1.462000e+01   
75%    5.186700e+02    643.000000   1594.380000   1414.555000  1.462000e+01   
max    5.186700e+02    644.530000   1616.910000   1441.490000  1.462000e+01   

       ...           s12           s13           s14           s15  \
count  ...  20631.000000  20631.000000  20631.000000  20631.000000   
mean   ...    521.413470   2388.096152   8143.752722      8.442146   
std    ...      0.737553      0.071919     19.076176      0.037505   
min    ...    518.690000   2387.880000   8099.940000      8.324900   
25%    ...    520.960000   2388.040000   8133.245000      8.414900   
50%    ...    521.480000   2388.090000   8140.540000      8.438900   
75%    ...    521.950000   2388.140000   8148.310000      8.465600   
max    ...    523.380000   2388.560000   8293.720000      8.584800   

                s16           s17      s18      s19           s20  \
count  2.063100e+04  20631.000000  20631.0  20631.0  20631.000000   
mean   3.000000e-02    393.210654   2388.0    100.0     38.816271   
std    1.556432e-14      1.548763      0.0      0.0      0.180746   
min    3.000000e-02    388.000000   2388.0    100.0     38.140000   
25%    3.000000e-02    392.000000   2388.0    100.0     38.700000   
50%    3.000000e-02    393.000000   2388.0    100.0     38.830000   
75%    3.000000e-02    394.000000   2388.0    100.0     38.950000   
max    3.000000e-02    400.000000   2388.0    100.0     39.430000   

                s21  
count  20631.000000  
mean      23.289705  
std        0.108251  
min       22.894200  
25%       23.221800  
50%       23.297900  
75%       23.366800  
max       23.618400  

[8 rows x 26 columns]

as we see there is 100 engine in the dataset with 362 cycles with mean of 108 cycles/ engine, last cycle in train data represent cycle which failure occured

## load test dataset

In [70]:
df_test = pd.read_csv('./data/PM_test.txt', sep=' ', header=None)
df_test.head()

0   1       2       3      4       5       6        7        8      9   \
0   1   1  0.0023  0.0003  100.0  518.67  643.02  1585.29  1398.21  14.62   
1   1   2 -0.0027 -0.0003  100.0  518.67  641.71  1588.45  1395.42  14.62   
2   1   3  0.0003  0.0001  100.0  518.67  642.46  1586.94  1401.34  14.62   
3   1   4  0.0042  0.0000  100.0  518.67  642.44  1584.12  1406.42  14.62   
4   1   5  0.0014  0.0000  100.0  518.67  642.51  1587.19  1401.92  14.62   

   ...       18      19    20   21    22     23     24       25  26  27  
0  ...  8125.55  8.4052  0.03  392  2388  100.0  38.86  23.3735 NaN NaN  
1  ...  8139.62  8.3803  0.03  393  2388  100.0  39.02  23.3916 NaN NaN  
2  ...  8130.10  8.4441  0.03  393  2388  100.0  39.08  23.4166 NaN NaN  
3  ...  8132.90  8.3917  0.03  391  2388  100.0  39.00  23.3737 NaN NaN  
4  ...  8129.54  8.4031  0.03  390  2388  100.0  38.99  23.4130 NaN NaN  

[5 rows x 28 columns]

In [71]:
df_test = prepare_df(df_test)
df_test.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29  1398.21   
1   1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45  1395.42   
2   1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94  1401.34   
3   1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12  1406.42   
4   1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19  1401.92   

      s5  ...     s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62  ...  521.72  2388.03  8125.55  8.4052  0.03  392  2388  100.0   
1  14.62  ...  522.16  2388.06  8139.62  8.3803  0.03  393  2388  100.0   
2  14.62  ...  521.97  2388.03  8130.10  8.4441  0.03  393  2388  100.0   
3  14.62  ...  521.38  2388.05  8132.90  8.3917  0.03  391  2388  100.0   
4  14.62  ...  522.15  2388.03  8129.54  8.4031  0.03  390  2388  100.0   

     s20      s21  
0  38.86  23.3735  
1  39.02  23.3916  
2  39.08  23.4166  
3  39.00  23.3737  
4  38.99  23.4130  

[5 rows x 26 columns]

In [29]:
df_test.shape

(13096, 26)

In [30]:
df_test.isna().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

In [31]:
df_test.describe()

id         cycle      setting1      setting2  setting3  \
count  13096.000000  13096.000000  13096.000000  13096.000000   13096.0   
mean      51.543907     76.836515     -0.000011      0.000004     100.0   
std       28.289423     53.057749      0.002203      0.000294       0.0   
min        1.000000      1.000000     -0.008200     -0.000600     100.0   
25%       28.000000     33.000000     -0.001500     -0.000200     100.0   
50%       52.000000     69.000000      0.000000      0.000000     100.0   
75%       76.000000    113.000000      0.001500      0.000300     100.0   
max      100.000000    303.000000      0.007800      0.000700     100.0   

                 s1            s2            s3            s4            s5  \
count  1.309600e+04  13096.000000  13096.000000  13096.000000  1.309600e+04   
mean   5.186700e+02    642.475088   1588.099204   1404.735362  1.462000e+01   
std    6.014263e-11      0.400899      5.003274      6.688309  2.671743e-12   
min    5.186700e+02    641.130000   1569.040000   1384.390000  1.462000e+01   
25%    5.186700e+02    642.197500   1584.600000   1399.950000  1.462000e+01   
50%    5.186700e+02    642.460000   1587.990000   1404.440000  1.462000e+01   
75%    5.186700e+02    642.740000   1591.362500   1409.050000  1.462000e+01   
max    5.186700e+02    644.300000   1607.550000   1433.360000  1.462000e+01   

       ...           s12           s13           s14           s15  \
count  ...  13096.000000  13096.000000  13096.000000  13096.000000   
mean   ...    521.747724   2388.071025   8138.947817      8.425844   
std    ...      0.559627      0.056934     10.188605      0.029009   
min    ...    519.380000   2387.890000   8108.500000      8.332800   
25%    ...    521.380000   2388.030000   8132.310000      8.405600   
50%    ...    521.780000   2388.070000   8138.390000      8.424900   
75%    ...    522.150000   2388.110000   8144.360000      8.444300   
max    ...    523.760000   2388.320000   8220.480000      8.541400   

                s16           s17      s18      s19           s20  \
count  1.309600e+04  13096.000000  13096.0  13096.0  13096.000000   
mean   3.000000e-02    392.571625   2388.0    100.0     38.892502   
std    8.823140e-15      1.233577      0.0      0.0      0.141681   
min    3.000000e-02    389.000000   2388.0    100.0     38.310000   
25%    3.000000e-02    392.000000   2388.0    100.0     38.800000   
50%    3.000000e-02    393.000000   2388.0    100.0     38.900000   
75%    3.000000e-02    393.000000   2388.0    100.0     38.990000   
max    3.000000e-02    397.000000   2388.0    100.0     39.410000   

                s21  
count  13096.000000  
mean      23.335743  
std        0.084120  
min       22.935400  
25%       23.281600  
50%       23.339200  
75%       23.393600  
max       23.641900  

[8 rows x 26 columns]

as we see that there is 100 engine, with 303 cycles with mean of 76 cycle/engine 

Same as training data, there are 100 engines, each engine has between 1 to 303 cycles (average of 76 cycles per engine). But this time, failure cycle was not provided.  

Failure events for test data - remaining cycles before failure (TTF) - were provided in a separate **truth file**.

## load truth dataset

In [72]:
df_truth = pd.read_csv('./data/PM_truth.txt', sep=' ', header=None)
df_truth.head()

0   1
0  112 NaN
1   98 NaN
2   69 NaN
3   82 NaN
4   91 NaN

In [73]:
df_truth.drop(columns=[1], axis=1, inplace=True)
df_truth.columns = ['ttf']
df_truth.head()

ttf
0  112
1   98
2   69
3   82
4   91

In [74]:
df_truth.shape

(100, 1)

In [36]:
df_truth.isna().sum()

ttf    0
dtype: int64

In [55]:
df_truth.describe()

ttf
count  100.00000
mean    75.52000
std     41.76497
min      7.00000
25%     32.75000
50%     86.00000
75%    112.25000
max    145.00000

here the is the time to failure (TTF) of each engine in the testing data. However, this column is not available in the test data. So, we need to add this column to the test data.

# feature engineering

* we will do some feature extraction and engineering to our dataset:
    - prepare data to regression by adding ttf label to train set
    - prepare data to classification by adding classifier label to train and test set for binary classification and multiclass classification

- regressiong answer the question: what is the expected time to failure (ttf) ?
- binary classification answer the question: is the machine going to fail in the a specific period of cycles ?
- multiclass classification answer the question: which period will be the most likely to fail ?

In [65]:
def add_feature(df_in, rolling_window):
    """
    args: 
        df_in: input dataframe
        rolling_window: rolling window period 
        
    returns: 
        df_out: output dataframe with added features
    """
    
    sensor_feature = ['s1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20', 's21']
    
    sensor_avg = [col.replace('s', 'avg') for col in sensor_feature]
    sensor_std = [col.replace('s', 'std') for col in sensor_feature]
    
    df_out = pd.DataFrame()
    rw = rolling_window
    
    for ids in df_in.id.unique():
        df_engine = df_in[df_in.id == ids]
        df_sub = df_engine[sensor_feature]
        
        # rolling mean
        av = df_sub.rolling(window=rw, min_periods=1).mean()
        av.columns = sensor_avg
        
        # rolling std
        std = df_sub.rolling(window=rw, min_periods=1).std().fillna(0)
        std.columns = sensor_std

        # concatinate dataframe
        df_new = pd.concat([df_engine, av, std], axis=1)
        df_out = pd.concat([df_out, df_new])
        
    return df_out

In [66]:
def prepare_train_df(df_in, period):
        
    """
    Add regression and classification labels to the training data.

        Regression label: ttf (time-to-failure) = each cycle for an engine subtracted from the last cycle of the same engine
        Binary classification label: label_bnc = if ttf is <= parameter period then 1 else 0. (values = 0,1)
        Multi-class classification label: label_mcc = 2 if ttf <= 0.5 * parameter period , 1 if ttf <= parameter period, else 0. (values = 0,1,2)
        
    Args:
        df_in (dataframe): The input training data
        period (int)     : The number of cycles for TTF segmentation. Used to derive classification labels
        
    Returns:
        dataframe: The input dataframe with regression and classification labels added

    
    illustrate the use of this function reset_index(level=0):     
    
                    Data
    Year Month      
    2000 Jan         1
        Feb         2
    2001 Jan         3
        Feb         4

    df.reset_index(level=0)

        Year  Data
    Month            
    Jan    2000     1
    Feb    2000     2
    Jan    2001     3
    Feb    2001     4
    """

    # regression label
    df_last_cycle = pd.DataFrame(df_in.groupby('id')['cycle'].max()).reset_index(level=0)
    df_last_cycle.columns = ['id', 'last_cycle']
    df_in = pd.merge(df_in, df_last_cycle, on='id')
    
    df_in['ttf'] = df_in['last_cycle'] - df_in['cycle']
    df_in = df_in.drop(columns=['last_cycle'], axis=1)
    
    # binary classification label
    df_in['bnc_label'] = df_in['ttf'].apply(lambda x: 1 if x <= period else 0) # values = 0,1. 0 = no failure, 1 = failure
    
    # multi-class classification label
    df_in['mcc_lable'] = df_in['ttf'].apply(lambda x: 2 if x <= 0.5 * period else 1 if x <= period else 0) # values = 0,1,2. 0 = no failure, 1 = low failure, 2 = high failure
    
    return df_in        

In [67]:
def prepare_test_df(df_test, df_truth, period):
    """
    Add regression and classification labels to the training data.

        Regression label: ttf (time-to-failure) = each cycle for an engine subtracted from the last cycle of the same engine
        Binary classification label: label_bnc = if ttf is <= parameter period then 1 else 0. (values = 0,1)
        Multi-class classification label: label_mcc = 2 if ttf <= 0.5 * parameter period , 1 if ttf <= parameter period, else 0. (values = 0,1,2)
        
    Args:
        df_test (dataframe): The input test data
        df_truth (dataframe): The input truth data  
        period (int)     : The number of cycles for TTF segmentation. Used to derive classification labels
        
    Returns:
        dataframe: The input dataframe with regression and classification labels added
    
    """    
    
    # regression label
    df_lst_test_cycle = pd.DataFrame(df_test.groupby('id')['cycle'].max())
    df_lst_test_cycle = df_lst_test_cycle.reset_index(level=0)
    df_lst_test_cycle.columns = ['id', 'last_cycle']
    df_test = pd.merge(df_test, df_lst_test_cycle, on='id')
    
    df_test = df_test[df_test['cycle'] == df_test['last_cycle']]
    df_test = df_test.reset_index(drop=True)
    df_test = df_test.drop(columns=['last_cycle'], axis=1)
    
    df = pd.concat([df_test, df_truth], axis=1)
    
    # binary classification label
    df['bnc_label'] = df['ttf'].apply(lambda x: 1 if x <= period else 0) # values = 0,1. 0 = no failure, 1 = failure
    
    # multi-class classification label
    df['mcc_label'] = df['ttf'].apply(lambda x: 2 if x <= 0.5 * period else 1 if x <= period else 0) # values = 0,1,2. 0 = no failure, 1 = low failure, 2 = high failure
    
    return df    

# prepare training data

In [44]:
df_train_prepared = add_feature(df_train, rolling_window=5)
df_train_prepared.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...     std12     std13     std14     std15  std16     std17  std18  \
0  14.62  ...  0.000000  0.000000  0.000000  0.000000    0.0  0.000000    0.0   
1  14.62  ...  0.438406  0.035355  5.041671  0.008697    0.0  0.000000    0.0   
2  14.62  ...  0.404475  0.026458  3.717450  0.007640    0.0  1.154701    0.0   
3  14.62  ...  0.495950  0.029439  3.050906  0.028117    0.0  1.000000    0.0   
4  14.62  ...  0.432574  0.025884  2.651326  0.025953    0.0  1.095445    0.0   

   std19     std20     std21  
0    0.0  0.000000  0.000000  
1    0.0  0.042426  0.003253  
2    0.0  0.055076  0.044573  
3    0.0  0.076322  0.037977  
4    0.0  0.073621  0.033498  

[5 rows x 68 columns]

In [49]:
df_train_prepared = prepare_train_df(df_train_prepared, period=30)
df_train_prepared.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...     std15  std16     std17  std18  std19     std20     std21  \
0  14.62  ...  0.000000    0.0  0.000000    0.0    0.0  0.000000  0.000000   
1  14.62  ...  0.008697    0.0  0.000000    0.0    0.0  0.042426  0.003253   
2  14.62  ...  0.007640    0.0  1.154701    0.0    0.0  0.055076  0.044573   
3  14.62  ...  0.028117    0.0  1.000000    0.0    0.0  0.076322  0.037977   
4  14.62  ...  0.025953    0.0  1.095445    0.0    0.0  0.073621  0.033498   

   ttf  bnc_label  mcc_lable  
0  191          0          0  
1  190          0          0  
2  189          0          0  
3  188          0          0  
4  187          0          0  

[5 rows x 71 columns]

In [50]:
df_train_prepared.shape

(20631, 71)

In [51]:
df_train_prepared.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20631 entries, 0 to 20630
Data columns (total 71 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         20631 non-null  int64  
 1   cycle      20631 non-null  int64  
 2   setting1   20631 non-null  float64
 3   setting2   20631 non-null  float64
 4   setting3   20631 non-null  float64
 5   s1         20631 non-null  float64
 6   s2         20631 non-null  float64
 7   s3         20631 non-null  float64
 8   s4         20631 non-null  float64
 9   s5         20631 non-null  float64
 10  s6         20631 non-null  float64
 11  s7         20631 non-null  float64
 12  s8         20631 non-null  float64
 13  s9         20631 non-null  float64
 14  s10        20631 non-null  float64
 15  s11        20631 non-null  float64
 16  s12        20631 non-null  float64
 17  s13        20631 non-null  float64
 18  s14        20631 non-null  float64
 19  s15        20631 non-null  float64
 20  s16   

In [52]:
df_train_prepared.isna().sum()

id           0
cycle        0
setting1     0
setting2     0
setting3     0
            ..
std20        0
std21        0
ttf          0
bnc_label    0
mcc_lable    0
Length: 71, dtype: int64

In [53]:
# save the prepared training data
df_train_prepared.to_csv('./data/df_train_prepared.csv', index=False)

# prepare test data

In [75]:
df_test_prepared = add_feature(df_test, rolling_window=5)
df_test_prepared.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29  1398.21   
1   1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45  1395.42   
2   1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94  1401.34   
3   1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12  1406.42   
4   1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19  1401.92   

      s5  ...     std12     std13     std14     std15  std16     std17  std18  \
0  14.62  ...  0.000000  0.000000  0.000000  0.000000    0.0  0.000000    0.0   
1  14.62  ...  0.311127  0.021213  9.948992  0.017607    0.0  0.707107    0.0   
2  14.62  ...  0.220681  0.017321  7.179807  0.032155    0.0  0.577350    0.0   
3  14.62  ...  0.337182  0.015000  5.890096  0.027781    0.0  0.957427    0.0   
4  14.62  ...  0.329742  0.014142  5.222300  0.024080    0.0  1.303840    0.0   

   std19     std20     std21  
0    0.0  0.000000  0.000000  
1    0.0  0.113137  0.012799  
2    0.0  0.113725  0.021642  
3    0.0  0.093095  0.020353  
4    0.0  0.080623  0.020672  

[5 rows x 68 columns]

In [76]:
df_test_prepared = prepare_test_df(df_test_prepared, df_truth, period=30)
df_test_prepared.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1     31   -0.0006    0.0004     100.0  518.67  642.58  1581.22  1398.91   
1   2     49    0.0018   -0.0001     100.0  518.67  642.55  1586.59  1410.83   
2   3    126   -0.0016    0.0004     100.0  518.67  642.88  1589.75  1418.89   
3   4    106    0.0012    0.0004     100.0  518.67  642.78  1594.53  1406.88   
4   5     98   -0.0013   -0.0004     100.0  518.67  642.27  1589.94  1419.36   

      s5  ...     std15  std16     std17  std18  std19     std20     std21  \
0  14.62  ...  0.021174    0.0  0.707107    0.0    0.0  0.124016  0.025037   
1  14.62  ...  0.011429    0.0  1.095445    0.0    0.0  0.069785  0.027820   
2  14.62  ...  0.030924    0.0  0.836660    0.0    0.0  0.143631  0.018966   
3  14.62  ...  0.021249    0.0  1.341641    0.0    0.0  0.126372  0.021106   
4  14.62  ...  0.009627    0.0  0.547723    0.0    0.0  0.091924  0.101100   

   ttf  bnc_label  mcc_label  
0  112          0          0  
1   98          0          0  
2   69          0          0  
3   82          0          0  
4   91          0          0  

[5 rows x 71 columns]

In [77]:
df_test_prepared.shape

(100, 71)

In [78]:
df_test_prepared.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 71 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         100 non-null    int64  
 1   cycle      100 non-null    int64  
 2   setting1   100 non-null    float64
 3   setting2   100 non-null    float64
 4   setting3   100 non-null    float64
 5   s1         100 non-null    float64
 6   s2         100 non-null    float64
 7   s3         100 non-null    float64
 8   s4         100 non-null    float64
 9   s5         100 non-null    float64
 10  s6         100 non-null    float64
 11  s7         100 non-null    float64
 12  s8         100 non-null    float64
 13  s9         100 non-null    float64
 14  s10        100 non-null    float64
 15  s11        100 non-null    float64
 16  s12        100 non-null    float64
 17  s13        100 non-null    float64
 18  s14        100 non-null    float64
 19  s15        100 non-null    float64
 20  s16        

In [79]:
df_test_prepared.isna().sum()

id           0
cycle        0
setting1     0
setting2     0
setting3     0
            ..
std20        0
std21        0
ttf          0
bnc_label    0
mcc_label    0
Length: 71, dtype: int64

In [80]:
# save the prepared test data
df_test_prepared.to_csv('./data/df_test_prepared.csv', index=False)